### ISSUE TITLE
Debugging help:  How can I trace an error that occurs before a model restart?

### ISSUE THREAD

**Issue by billdenney** [nlmixr2/nlmixr2est#375]
I have a model where I get the following error:

```
Error in names(object) <- nm : 
  'names' attribute [14] must be the same length as the vector [0]
```

and, after the error, I see "Restart 1" (or 2 or 3).  I'd like to trace where that error occurs, but my normal method of `options(error = recover)` doesn't work because it's not an error (or I think more accurately, it's captured internally).  Is there a way that I can trace the error to its source?

**Comment by billdenney** [comment]
Thanks, that helped.

My error is coming from here: https://github.com/nlmixr2/nlmixr2est/blob/74a732ee44d8bfbea4cedde5961aa334854aca3d/R/focei.R#L1397

The error message from `foceiFitCpp_()` appears to be `"theta reset0"` which seems to indicate a zero gradient.

Looking more into the code, my best guess is that `thetaReset` is not being assigned within `inner.cpp`.  My *guess* is that `doAdjust` is not set to  `true` before here, so it's never calling `thetaReset00` below here:

https://github.com/nlmixr2/nlmixr2est/blob/74a732ee44d8bfbea4cedde5961aa334854aca3d/src/inner.cpp#L1988

Any insight into how I could troubleshoot more (realizing that I can't share the model/data itself)?

P.S.  I may be able to make a reproducible example, eventually because the underlying issue is a parameter that is not informed by any of the data.

**Comment by billdenney** [comment]
P.P.S.  My attempt to replicate the error in a model I could share gave a different zero gradient error.  I don't know of a simple replication method.

**Comment by mattfidler** [comment]
This is probably the hardest part to debug, but perhaps I can do something...

**Comment by billdenney** [comment]
Is there a sprinkling of `printf`s that may help (without being too much noise in the inner loop)?

**Comment by mattfidler** [comment]
You could add where it is wherever the `thetaReset` is called, but you would need to use something like `REprintf()` instead of `printf`.

I can't simply look at the code and figure out what is going on in your case.

**Comment by billdenney** [comment]
I just changed the function so that it  moved the `doAdjust` test to the bottom of the function.  Specifically, I moved this line

https://github.com/nlmixr2/nlmixr2est/blob/74a732ee44d8bfbea4cedde5961aa334854aca3d/src/inner.cpp#L1988

To just above this line

https://github.com/nlmixr2/nlmixr2est/blob/74a732ee44d8bfbea4cedde5961aa334854aca3d/src/inner.cpp#L2009

Then, I got the following message:

```
Theta reset (zero/bad gradient values); Switch to bobyqa
```

That seems like the desired behavior when there is a zero derivative, but maybe I'm missing something important.  Would that be a good fix?  It does appear to be related to not calling the `thetaReset00()` function, but I don't know what is happening in that part of the code well enough to be sure that it's a good fix.

**Comment by mattfidler** [comment]
Well if there is a zero gradient then yes that is a good fix.

But, the adjust readjusts the parameters to the correct scale sometimes changing from an eta drift to modifying the theta and then continuing the estimation

**Comment by billdenney** [comment]
(TL;DR: Should I move line 1988 to just before 2009 as a PR for this?  I'm not sure.)

My real issue with this model is that there is a zero gradient that does not appear until after the initial estimation; I don't know exactly the cause (I would have thought that one of the parameters would have a zero gradient on the first estimation step), so I can't readily make a reproducible example that I can share.

The "But..." part makes me wonder/worry that I shouldn't make the suggested fix.  It looks like the code after line 1988 is possibly shifting the etas around, and then `thetaReset00()` is called to send those results back out to R.

I don't think that doing the work of the lines after 1988 would be a problem, but I'm also not sure if I may be missing something.

**Comment by mattfidler** [comment]
I don't think that is the right solution.  Without the adjustment you may not continue the optimization at the same point.

**Comment by mattfidler** [comment]
Could try #384

**Comment by mattfidler** [comment]
This has been resolved.

The underlying bug behind the `'names' attribute [14] must be the same length as the vector [0]` error was that, on a zero/bad gradient, the C++ `thetaResetZero()` -> `thetaReset0()` path could bail out early (when `doAdjust` was `false`) **without** ever calling `thetaReset00()`. Since `thetaReset00()` is what populates the `.thetaReset` environment (`thetaIni`, `etaMat`, `omegaTheta`, ...), the R-side restart then ran `setNames(.thetaReset$thetaIni + 0.0, .nm)` against a length-0 vector, producing that error.

Rather than moving the `doAdjust` check (which would have skipped the parameter re-scaling), #384 forces the reset while keeping the adjustment: `thetaResetZero()` now calls `thetaReset0(true)`, so `thetaReset00()` always runs and the `.thetaReset` environment is populated before the restart. This has been in every release since v2.2.0.

The zero-gradient handling was also hardened afterwards with dedicated, default-`TRUE` control options in `foceiControl()`: `zeroGradFirstReset`, `zeroGradRunReset`, and `zeroGradBobyqa` (the last being the "Switch to bobyqa" behavior you saw and liked).

Closing as resolved -- please reopen if you still hit this on a current version.
